src/data/zmod/basic.lean

```diff
@@ -577,6 +577,79 @@ def units_equiv_coprime {n : ℕ} [fact (0 < n)] :
   left_inv := λ ⟨_, _, _, _⟩, units.ext (nat_cast_zmod_val _),
   right_inv := λ ⟨_, _⟩, by simp }
 
+/-- A ring `R` isomorphic to `R × zmod 1` -/
+def prod_zmod_one_equiv (R : Type*) [semiring R] : R ≃+* R × zmod 1 :=
+{ to_fun := λ x, (x, 0),
+  inv_fun := prod.fst,
+  map_add' := by intros; ext; simp; congr,
+  map_mul' := by intros; ext; simp; congr,
+  left_inv := λ x, rfl,
+  right_inv := λ x, by apply prod.ext; simp; congr }
+
+@[simp] lemma prod_zmod_one_equiv_apply {R : Type*} [semiring R] (x : R) :
+  prod_zmod_one_equiv R x = (x, 0) := rfl
+
+@[simp] lemma prod_zmod_one_equiv_symm_apply {R : Type*} [semiring R] (x : R × zmod 1) :
+  (prod_zmod_one_equiv R).symm x = x.fst := rfl
+
+/-- A ring `R` isomorphic to `zmod 1 × R` -/
+def zmod_one_prod_equiv (R : Type*) [semiring R] : R ≃+* zmod 1 × R :=
+{ to_fun := λ x, (0, x),
+  inv_fun := prod.snd,
+  map_add' := by intros; ext; simp; congr,
+  map_mul' := by intros; ext; simp; congr,
+  left_inv := λ x, rfl,
+  right_inv := λ x, by apply prod.ext; simp; congr }
+
+@[simp] lemma zmod_one_prod_equiv_apply {R : Type*} [semiring R] (x : R) :
+  zmod_one_prod_equiv R x = (0, x) := rfl
+
+@[simp] lemma zmod_one_prod_equiv_symm_apply {R : Type*} [semiring R] (x : zmod 1 × R) :
+  (zmod_one_prod_equiv R).symm x = x.snd := rfl
+
+/-- The **Chinese remainder theorem**. For a pair of coprime natural numbers, `m` and `n`,
+  the rings `zmod (m * n)` and `zmod m × zmod n` are isomorphic. -/
+def chinese_remainder {m n : ℕ} (h : m.coprime n) :
+  zmod (m * n) ≃+* zmod m × zmod n :=
+if hmn0 : m * n = 0
+then if hm1 : m = 1
+  then by rw [hm1, one_mul]; exact zmod_one_prod_equiv _
+  else have hn1 : n = 1,
+      begin
+        rw [nat.mul_eq_zero] at hmn0,
+        rcases hmn0 with ⟨rfl, rfl⟩;
+        simp * at *
+      end,
+    by rw [hn1, mul_one]; exact prod_zmod_one_equiv _
+else
+let to_fun : zmod (m * n) → zmod m × zmod n :=
+  λ x, (zmod.cast_hom (dvd_mul_right _ _) _ x, zmod.cast_hom (dvd_mul_left _ _) _ x) in
+let inv_fun : zmod m × zmod n → zmod (m * n) :=
+  λ x, nat.modeq.chinese_remainder h x.1.val x.2.val in
```

Comment:
This bears a passing resemblence to `fin_prod_fin_equiv` - can they be linked together in any way?